### DISCUSSION TITLE
How to Disable Buffering for Real-Time Transform Data from NDI Tracker in PlusServer?

### DISCUSSION THREAD

**Discussion by ArashDadras** [PlusToolkit/PlusLib#1203] [Q&A]
Hi everyone.
I'm new to PlusServer and I'm learning to use it for a project where I need to get transform data from NDI Vicra\Vega in real time and use it in Unity. I can get the data correctly from PlusServer but it seems that I am getting the transform data that belongs to miliseconds or seconds ago. Is there a buffering mechanism in PlusServer or something else causing this? If there is buffering, can I disable it to get only the most recent data?
 
The config I use to run the server:
`<PlusConfiguration version="2.8">
  <DataCollection StartupDelaySec="1.0" >
    <DeviceSet
      Name="PlusServer: Vicra Vega tracker with passive markers - by IP"
      Description="Broadcasting tool tracking data through OpenIGTLink" />
    <Device
      Id="TrackerDevice"
      Type="NDITracker"
      SerialPort="5"
      BaudRate="921600"
      AcquisitionRate="200"
      ToolReferenceFrame="Tracker" >	
      <DataSources>
		  <DataSource 
			  Type="Tool" 
			  Id="Reference" 	
			  RomFile="D:/References/HeadBand-Reference/HeadBand-Reference.rom"  
		   />
		  <DataSource Type="Tool" Id="Pointer" 	RomFile="D:/References/Navigation-Pointer/Navigation-Pointer.rom" />
      </DataSources>
      <OutputChannels>
        <OutputChannel Id="TrackerStream" >
          <DataSource Id="Reference"/>
          <DataSource Id="Pointer"/>
        </OutputChannel> 
      </OutputChannels>
    </Device>
  </DataCollection>
  <PlusOpenIGTLinkServer
    ListeningPort="18944"
    SendValidTransformsOnly="true"
    MaxNumberOfIgtlMessagesToSend="10"
    MaxTimeSpentWithProcessingMs="10"
    OutputChannelId="TrackerStream" >
    <DefaultClientInfo>
      <MessageTypes>
        <Message Type="TRANSFORM" />
      </MessageTypes>
      <TransformNames>
		<Transform Name="ReferenceToTracker" />
		<Transform Name="PointerToTracker" />
	  </TransformNames>
    </DefaultClientInfo>
  </PlusOpenIGTLinkServer>
</PlusConfiguration>`

**Comment by lassoan** [comment]
Few-ten-millisecond delay is possible, but a if you experience a one-second delay then something is wrong.

If you are slow to receive and process messages then it may indeed increase latency, because Plus tries to send all acquired data without data loss. If you cannot speed up your receiver then your can adjust the server parameters: set `MaxNumberOfIgtlMessagesToSend="1" `, maybe reduce `MaxTimeSpentWithProcessingMs`. Make sure that there is no excessive logging (it is expensive and nasty show things down).

**Reply by ArashDadras** [reply]
Thanks for the replay. I have changed these parameters in the configuration but I still have a delay. I'm getting the position of the tool's in 5 to 10 frames before instead of its current position. In the documents I came across LocalTimeOffsetSec, does this apply to my problem?

**Reply by lassoan** [reply]
LocalTimeOffsetSec only used for synchronizing multiple data streams. How do you know that you are getting the tool position from 5-10 frames before?

You can also try to reduce aquisition frame rate to reduce data congestion.

**Reply by ArashDadras** [reply]
Sorry I had to clarify more, I'm developing an AR app with unity which I need to place a .fbx model on the position of tracked tools. I can align a 3D model with the tool but the problem is when the tool is moved the 3d model follows the path of the tool and after some time (usually less than a second) reaches the tool. That's why I say I have a 5-10 frame delay. I've also measured the timestamps of the transformation that comes from plusttoolkit with system time, which gives us about 300 to 500 msc delay. I understand that this problem may not be related to PlusToolKit, but I want to be sure that my configuration is not wrong and causing the delay. I will try reducing the aquisition frame rate parameter.
